core/amber/src/main/scala/edu/uci/ics/texera/web/resource/ResultDownloadResource.scala

```diff
@@ -0,0 +1,202 @@
+package edu.uci.ics.texera.web.resource
+
+import java.util
+
+import com.google.api.client.util.Lists
+import com.google.api.services.drive.Drive
+import com.google.api.services.drive.model.Permission
+import com.google.api.services.sheets.v4.Sheets
+import com.google.api.services.sheets.v4.model.{
+  AppendValuesResponse,
+  Spreadsheet,
+  SpreadsheetProperties,
+  ValueRange
+}
+import edu.uci.ics.amber.engine.common.tuple.ITuple
+import edu.uci.ics.texera.web.model.event.ResultDownloadResponse
+import edu.uci.ics.texera.web.model.request.ResultDownloadRequest
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import javax.websocket.Session
+
+object ResultDownloadResource {
+
+  private final val SESSION_CACHED_KEY_NAME = "ResultDownloadResource"
+
+  private final val UPLOAD_SIZE = 100;
+
+  def clearCache(session: Session): Unit = {
+    session.getUserProperties
+      .entrySet()
+      .removeIf(entry => entry.getKey.startsWith(SESSION_CACHED_KEY_NAME))
+  }
+
+  def apply(
+      session: Session,
+      request: ResultDownloadRequest,
+      sessionResults: Map[String, List[ITuple]]
+  ): ResultDownloadResponse = {
+    // look for the response cached in the session to avoid creating duplicated file
+    if (
+      session.getUserProperties.containsKey(
+        getSessionName(request.downloadType)
+      )
+    ) {
+      return session.getUserProperties
+        .get(getSessionName(request.downloadType))
+        .asInstanceOf[ResultDownloadResponse]
+    }
+
+    // By now the workflow should finish running. Only one operator should contain results.
+    val count = sessionResults.count(p => !p._2.isEmpty)
+    if (count == 0) {
+      return ResultDownloadResponse(
+        request.downloadType,
+        "",
+        "The workflow contains no results"
+      )
+    } else if (count > 1) {
+      return ResultDownloadResponse(
+        request.downloadType,
+        "",
+        "The workflow does not finish running"
+      )
+    }
```

Comment:
it is not clear to me what does `count` mean. why does it considered not finished running if `count>1`? please add more comments or make the code more readable.